tests/README.md

```diff
@@ -0,0 +1,1510 @@
+		$this->assertInstanceOf
+					( Shy::class
+					, $this->subject->shy("label", "http://www.ilias.de")
+					);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_close_button() {
+		$this->assertInstanceOf
+					( Close::class
+					, $this->subject->close("label", "http://www.ilias.de")
+					);
+	}
+}
+```
+
+
+<a name="useless-test--generic-naming"></a>
+#### Useless test / Generic naming
+Unit tests should always have an assertion of the result, because of that PHPUnit 6 started to mark such
+tests as useless. Useless test are always threaded as failed. Furthermore, the test name *test_button_label_or_glyph_only*
+is not really telling whats exactly tested.
+
+Another aspect of the whole *ButtonTest* class is that a factory is used to create concrete instances of
+the Buttons. But the factory nor the concrete subclasses of the Button class is a test subject here only the
+Button class itself. Of course to test all button instances is not really effective. However, the button can
+also be created with the help of mockery which subclasses the button dynamically within the tests. This allows
+to test the Button class in a dedicated way. If logic is added to one of the specific implementations only
+this part has to be tested within the test class of the specific button subclass for example the primary button.
+
+
+```php
+<?php
+class ButtonTest extends ILIAS_UI_TestBase {
+
+	public function getButtonFactory() {
+			return new \ILIAS\UI\Implementation\Component\Button\Factory();
+	}
+
+	public function button_type_provider() {
+			return array
+				( array("standard")
+				, array("primary")
+				, array("shy")
+				, array("tag")
+				);
+	}
+
+	/**
+	 * @dataProvider button_type_provider
+	 */
+	public function test_button_label_or_glyph_only($factory_method) {
+		$f = $this->getButtonFactory();
+		try {
+			$f->$factory_method($this, "http://www.ilias.de");
+			$this->assertFalse("This should not happen");
+		}
+		catch (\InvalidArgumentException $e) {}
+	}
+}
+```
+
+<a name="example-solution-1"></a>
+##### Example solution
+The proposed solution of this example would be to remove the test because the trait should be tested in a separate
+class dedicated to the trait. However, if the author decides to test the class which use the trait in stead of the trait itself
+the solution would look like this.
+
+First the *Factory* was removed because this class is meant for the *Button* class.
+Second the data provider has been removed due to the fact that the button itself will be tested and not
+the children of the button.
+Third the try catch was replaced with the phpunit construct which is designed to test exception occurrence.
+Finally the *Button* class has been partial mocked with a full method delegation which means that the test code
+is directly talking to the real button implementation. The second parameter is a list of construct argument for the button
+which remained unchanged to get same result as before.
```

Comment:
i.O. possible to do it this way. Other tests may be outsourced to respective other test classes. However, this seems to be a lot of overhead (we have to repeat this test in all other specific classes). What do the others think of this?

Reply:
The fundamental missunderstanding seems to be that the button-class is under test here. It's not. There of course is room for improvement.

In general imo there are other very reasonable definitions of "unit" besides "class". This depends on what level of abstraction the tests are written for. After all, the concrete classes could be considered an implementation detail for a lot of cases. Instead of testing a spec for an abstraction in my unit tests I start to test specifics of the implementation I came up with to deliver the desired behaviour of the abstraction. But these are different things.  When tests rely to much on a concrete implementation, they become worthless as well because they need to change when implementation changes.

Write testcase classes for classes is a good rule of thumb but fails to be generally correct. Sometimes a single class is just not a very interesting thing. Sometimes a single class might be to large to be considered a unit. Not all automated tests are unit tests and even though it's PHP_Unit_ I surely could use the exact same lib to write other kinds of automated tests. Sometimes I even doubt the seperation between "unit test", "integration test", "smoke test" and what not is valuable. I need to know where I could draw a line around stuff in my system, what behaviour I want at that boundary and how I control that thingy inside to test that behaviour. To call it "unit test" or "integration test" then is just a matter of scope.

Reply:
Of course you are right @klees there are other definitions of unit like the Result you talked about in an other part of #2. However I still a bit cautious about this units because there are a lot of tests which make sense if the "unit" is tested but the subparts or "code paths" are never failing in such a way, for example the result->isOk as a test sounds ok but success->isOk is not really interesting. However only success->isOk or failure->isOk could actually happen which leads to the question if such a unit is really the optimal way to go.

In my opinion the test is the specification of the implementation, therefore if a class changes its behaviour the test / specification has to change as well. Of course the tests should never rely on an internal part of the implementation but on its public interface.
Thats why I can't full agree with your opinion that test become worthless if they have to change with the implementation of the class.

I fully agree that you are able to write integration tests with PHPUnit due to the [xUnit](https://en.wikipedia.org/wiki/XUnit) architecture. I guess thats why it is called PHPUnit. There are also optional packages which are used to add seeds to the db for example. 

The separation of test scopes exist because they serve different purposes. The unit test is a specification of the smallest testable unit where ever this line is, may its a class or multiple classes which form a logical unit. Integration tests are verifying the integration of the units with each other. End to End tests check if one feature slice with GUI works as expected and so on. However you are right, in some cases are the lines a bit grayish :) 

It seemed to me that you not really agreed with the proposed solution which is not a bad thing at all, how would you like to improve the proposed solution?

@Amstutz I am not really sure if I got you right, but you don't need to outsource the test methods unless they are specific for that button class.

Reply:
I didn't say, that tests are worthless if they need to change with the implementation sometimes. I merely said, that the tests should not rely too much on the concrete implementation, because they would need to change when that implementation changes. That's a huge difference. I also think we agree on that point, because after all we want to test the behaviour (i.e. public interface) of an implementation, not specifics. So changes in behaviour should introduce (or be introduced by) changes in tests, while e.g. switching a `foreach` for a `while` for refactoring purpose (or similar reasons) should not change the test if no change of behaviour is desired.

After all, I write my testcases according to the specifications I have for my unit. I intentionally do not factor in any assumptions about which paths through the code might fail in the beginning. I might do that when I encounter specific intricacies or bugs during the implementation or even later, but I might have added a testcase for the concrete implementation then, not for the specification of the public interface. So the interesting thing in the beginning imo is "does the unit follow specifications" and not "does a thing fail which I expect to go wrong".

I understand why I might want to have different names for tests of different purposes, but In the end tests need to give trust that something works according to a specification, may it be a unit, a system of different units or a system of units with some GUI in front. From the perspective of the GUI, I don't care of how many units the developer constructed it and if these are under tests on their own. Specs need to be followed. I if use a subsystem as a developer, I do not care of how many units this subsystem might consist. But: Specs need to be followed. There even might be unit, that starts as a single unit, but will be later deconstructed in separate parts. It might well happen, that the actual specifications of that thingy and the tests for them do not change in that process. So "smallest testable part" could be understood as "smallest part I have specifications for".

In the concrete case in question here: There is no spec that says "Buttons should derive from a common base class" or "The UI-framework needs to implement its distinct components by distinct classes" or "there is a button class that rejects non-labels or non-glyphs as argument". There indeed is a spec that says "We may only have a label or a glyph on a button.". That is the thing that was under test here.

By your proposed improvement, the conclusion of the test changed a lot. Instead of testing "every type of button may only have a label or glyph on it" you are testing "the button class rejects invalid arguments on construction". There is no such thing as a button class in the public interface of the UI service. There indeed is factory, that allows me to construct buttons. So maybe from you POV that test can't be specified as a unit test, but it indeed does test a spec, and imo properly so.